programs/stake-tests/tests/test_move_stake_and_lamports.rs

```diff
@@ -0,0 +1,1281 @@
+    instruction: &Instruction,
+    additional_signers: &T,
+) -> ProgramResult {
+    let mut transaction =
+        Transaction::new_with_payer(&[instruction.clone()], Some(&context.payer.pubkey()));
+
+    transaction.partial_sign(&[&context.payer], context.last_blockhash);
+    transaction.sign(additional_signers, context.last_blockhash);
+
+    match context.banks_client.process_transaction(transaction).await {
+        Ok(_) => Ok(()),
+        Err(e) => {
+            // banks client error -> transaction error -> instruction error -> program error
+            match e.unwrap() {
+                TransactionError::InstructionError(_, e) => Err(e.try_into().unwrap()),
+                TransactionError::InsufficientFundsForRent { .. } => {
+                    Err(ProgramError::InsufficientFunds)
+                }
+                _ => panic!("couldnt convert {:?} to ProgramError", e),
+            }
+        }
+    }
+}
+
+async fn test_instruction_with_missing_signers(
+    context: &mut ProgramTestContext,
+    instruction: &Instruction,
+    additional_signers: &Vec<&Keypair>,
+) {
+    // remove every signer one by one and ensure we always fail
+    for i in 0..instruction.accounts.len() {
+        if instruction.accounts[i].is_signer {
+            let mut instruction = instruction.clone();
+            instruction.accounts[i].is_signer = false;
+            let reduced_signers: Vec<_> = additional_signers
+                .iter()
+                .filter(|s| s.pubkey() != instruction.accounts[i].pubkey)
+                .collect();
+
+            let e = process_instruction(context, &instruction, &reduced_signers)
+                .await
+                .unwrap_err();
+            assert_eq!(e, ProgramError::MissingRequiredSignature);
+        }
+    }
+
+    // now make sure the instruction succeeds
+    process_instruction(context, instruction, additional_signers)
+        .await
+        .unwrap();
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+enum StakeLifecycle {
+    Uninitialized = 0,
+    Initialized,
+    Activating,
+    Active,
+    Deactivating,
+    Deactive,
+}
+impl StakeLifecycle {
+    // (stake, staker, withdrawer)
+    async fn new_stake_account(
+        self,
+        context: &mut ProgramTestContext,
+        vote_account: &Pubkey,
+        staked_amount: u64,
+    ) -> (Keypair, Keypair, Keypair) {
+        let stake_keypair = Keypair::new();
+        let staker_keypair = Keypair::new();
+        let withdrawer_keypair = Keypair::new();
+
+        self.new_stake_account_fully_specified(
+            context,
+            vote_account,
+            staked_amount,
+            &stake_keypair,
+            &staker_keypair,
+            &withdrawer_keypair,
+            &Lockup::default(),
+        )
+        .await;
+
+        (stake_keypair, staker_keypair, withdrawer_keypair)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    async fn new_stake_account_fully_specified(
```

Comment:
yea i wanted to use the actual stake program mechanism to set accounts up rather than hardcoded data because its more likely to catch regressions/changes. i think i can do better than this tho, a function that can create multiple accounts at once means i could also move the logic to fix the epochs into it